langchain/src/callbacks/handlers/tracers.ts

```diff
@@ -264,8 +273,48 @@ export abstract class BaseTracer extends BaseCallbackHandler {
     const toolRun = run as ToolRun;
     toolRun.end_time = Date.now();
     toolRun.error = error.message;
+    await this.onToolError?.(toolRun);
     await this._endTrace(toolRun);
   }
+
+  async handleAgentAction(action: AgentAction, runId: string): Promise<void> {
+    const run = this.runMap.get(runId);
+    if (!run || run?.type !== "chain") {
+      return;
+    }
+    const agentRun = run as AgentRun;
+    agentRun.actions = agentRun.actions || [];
+    agentRun.actions.push(action);
+    await this.onAgentAction?.(run as AgentRun);
+  }
+
+  // custom event handlers
+
+  onLLMStart?(run: LLMRun): void | Promise<void>;
```

Comment:
@agola11 I've added these separate methods because
1. It's much nicer to work with the Run objects
2. It's tricky to know exactly when to call them, eg. in start is in the last line, in end methods it's right before we delete the run etc

Reply:
looks reasonable. Do we really need the onAgentRun/handleText here though?